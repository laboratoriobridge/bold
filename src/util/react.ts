import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
