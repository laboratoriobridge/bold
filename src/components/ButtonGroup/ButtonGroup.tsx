import React, { CSSProperties } from 'react'

import { ExternalStyles, useStyles } from '../../styles'
import { Omit } from '../../util'

export interface ButtonGroupProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  style?: ExternalStyles
}

export function ButtonGroup(props: ButtonGroupProps) {
  const { style, ...rest } = props
  const { classes, css } = useStyles(createStyles)

  return <div className={css(classes.root, style)} role='group' {...rest} />
}

export const createStyles = () => ({
  root: {
    '& > button:not(:first-of-type), & > a:not(:first-of-type)': {
      borderTopLeftRadius: '0',
      borderBottomLeftRadius: '0',
    },
    '& > button:not(:last-of-type), & > a:not(:last-of-type)': {
      borderTopRightRadius: '0',
      borderBottomRightRadius: '0',
    },
    '& > button:focus, & > a:focus': {
      zIndex: 1, // prevent box-shadow overlapping
    },
  } as CSSProperties,
})
