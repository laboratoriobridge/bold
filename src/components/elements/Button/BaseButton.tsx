import * as React from 'react'

export type OnClickWithPromise = (event: React.MouseEvent<any>) => any

export interface BaseButtonProps {
    /**
     * css className
     */
    className?: string
    disabled?: boolean
    name?: string
    onClick?: React.MouseEventHandler<any> | OnClickWithPromise
    onMouseEnter?: React.MouseEventHandler<any>
    onMouseLeave?: React.MouseEventHandler<any>
    tabIndex?: number
    title?: string
    innerRef?: React.Ref<HTMLElement>
    onLoadingChange?(loading: boolean): void
    render?(props: any): React.ReactNode
}

export class BaseButton extends React.Component<BaseButtonProps> {

    static defaultProps: BaseButtonProps = {
        render: ({ innerRef, ...rest }: any) => {
            return (
                <button type='button' ref={innerRef} {...rest} />
            )
        },
    }

    private timeout: number

    componentWillUnmount() {
        clearTimeout(this.timeout)
        this.timeout = -1
    }

    render() {
        const { onLoadingChange, render, ...rest } = this.props

        return render({
            ...rest,
            onClick: this.onClick,
            onKeyPress: this.handleOnKeyPress,
        })
    }

    private handleOnKeyPress = (event) => {
        event.preventDefault()
        if (!event) { event = window.event } // cross-browser shenanigans
        if (event.charCode === 32 || event.charCode === 13 && this.props.onClick) { // this is the spacebar
            this.onClick(event)
        }
        return true // treat all other keys normally;
    }

    private onClick = (event: React.MouseEvent<any>) => {
        if (this.props.onClick) {

            const promise = this.props.onClick(event)
            if (promise && promise.then) {
                this.startLoading()
                promise
                    .then(() => this.stopLoading())
                    .catch((error) => {
                        this.stopLoading()
                        throw new Error(error)
                    })
            }
        }
    }

    private startLoading = () => {
        this.props.onLoadingChange && this.props.onLoadingChange(true)
    }

    private stopLoading() {
        if (!this.timeout) {
            this.timeout = window.setTimeout(() => {
                this.props.onLoadingChange && this.props.onLoadingChange(false)
                this.timeout = null
            }, 10)
        }
    }

}
