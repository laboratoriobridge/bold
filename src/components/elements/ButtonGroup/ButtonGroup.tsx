import { Interpolation } from 'emotion'
import * as React from 'react'

import { Styles, withStyles, WithStylesProps } from '../../../styles/withStyles'

export interface ButtonGroupProps extends WithStylesProps {
    style?: Interpolation
}

@withStyles
export class ButtonGroup extends React.PureComponent<ButtonGroupProps> {
    render() {
        const { children, css, style } = this.props
        const styles: Styles = {
            root: {
                '& > button:not(:first-of-type), & > a:not(:first-of-type)': {
                    borderTopLeftRadius: '0',
                    borderBottomLeftRadius: '0',
                },
                '& > button:not(:last-of-type), & > a:not(:last-of-type)': {
                    borderTopRightRadius: '0',
                    borderBottomRightRadius: '0',
                },
                '& > button:focus, & > a:focus': {
                    zIndex: 1, // prevent box-shadow overlapping
                },
            },
        }

        return (
            <div className={css(styles.root, style)} role='group'>
                {children}
            </div>
        )
    }
}
