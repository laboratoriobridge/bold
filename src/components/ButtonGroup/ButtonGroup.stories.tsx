import React from 'react'

import { Button } from '../Button'
import { HFlow } from '../HFlow'
import { Icon } from '../Icon'

import { ButtonGroup } from './ButtonGroup'

export default {
  title: 'Components/ButtonGroup',
}

export const Default = () => (
  <HFlow>
    <ButtonGroup>
      <Button size='small' kind='primary'>
        First
      </Button>
      <Button size='small'>Second</Button>
      <Button size='small' disabled>
        Third
      </Button>
      <Button size='small' disabled>
        Fourth
      </Button>
    </ButtonGroup>
    <ButtonGroup>
      <Button size='small'>
        <Icon icon='trashOutline' />
      </Button>
      <Button size='small'>
        <Icon icon='zoomOutline' />
      </Button>
    </ButtonGroup>
  </HFlow>
)
