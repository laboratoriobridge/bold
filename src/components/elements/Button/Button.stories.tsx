import { action } from '@storybook/addon-actions'
import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'
import { MemoryRouter } from 'react-router'

import { Icon } from '../Icon'
import { Text } from '../textual'

import { Button } from './Button/Button'
import { ButtonKind, ButtonSize, ButtonSkin } from './Button/ButtonSkins'
import { ButtonLink } from './ButtonLink/ButtonLink'

const typeOptions: ButtonKind[] = ['normal', 'primary', 'danger']
const skinOptions: ButtonSkin[] = ['default', 'ghost', 'outline']
const sizeOptions: ButtonSize[] = ['small', 'medium', 'large']

storiesOf('Components|Button', module)
  .add('default', () => (
    <Button
      skin={select('skin', skinOptions, 'default')}
      size={select('size', sizeOptions, 'medium')}
      kind={select('kind', typeOptions, 'normal')}
      loading={boolean('loading', false)}
      disabled={boolean('disabled', false)}
      block={boolean('block', false)}
      onClick={action('button-clicked')}
    >
      {text('children', 'Button')}
    </Button>
  ))
  .add('icon', () => (
    <Button
      skin={select('skin', skinOptions, 'ghost')}
      size={select('size', sizeOptions, 'small')}
      kind={select('kind', typeOptions, 'normal')}
      loading={boolean('loading', false)}
      disabled={boolean('disabled', false)}
      block={boolean('block', false)}
      onClick={action('button-clicked')}
    >
      <Icon icon='adjust' />
    </Button>
  ))
  .add('link', () => (
    <MemoryRouter initialEntries={['/']} initialIndex={0}>
      <ButtonLink
        to='/links'
        skin={select('skin', skinOptions, 'default')}
        size={select('size', sizeOptions, 'medium')}
        kind={select('kind', typeOptions, 'normal')}
        loading={boolean('loading', false)}
        disabled={boolean('disabled', false)}
        block={boolean('block', false)}
        onClick={action('link-clicked')}
      >
        <Icon icon='arrowLeft' style={{ marginRight: '0.5rem' }} />
        <Text>Go back</Text>
      </ButtonLink>
    </MemoryRouter>
  ))
