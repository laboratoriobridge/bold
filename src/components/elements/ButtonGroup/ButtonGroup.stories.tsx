import { storiesOf } from '@storybook/react'
import React from 'react'

import { Button } from '../Button'
import { HFlow } from '../Flow'
import { Icon } from '../Icon'

import { ButtonGroup } from './ButtonGroup'

storiesOf('Components|ButtonGroup', module).add('default', () => (
  <HFlow>
    <ButtonGroup>
      <Button size='small' kind='primary'>
        First
      </Button>
      <Button size='small'>Second</Button>
      <Button size='small' disabled>
        Third
      </Button>
    </ButtonGroup>
    <ButtonGroup>
      <Button size='small'>
        <Icon icon='trashOutline' />
      </Button>
      <Button size='small'>
        <Icon icon='zoomOutline' />
      </Button>
    </ButtonGroup>
  </HFlow>
))
