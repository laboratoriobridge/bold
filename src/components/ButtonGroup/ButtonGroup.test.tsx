import { render } from '@testing-library/react'
import React from 'react'

import { Button } from '../Button'

import { ButtonGroup } from './ButtonGroup'

it('should render correctly', () => {
  const { container } = render(
    <ButtonGroup>
      <Button>First</Button>
    </ButtonGroup>
  )
  expect(container).toMatchSnapshot()
})

it('should should accept style prop', () => {
  const { container } = render(<ButtonGroup style={{ color: 'red' }} />)
  expect(container).toMatchSnapshot()
})

it('should should accept HTML div element props', () => {
  const { container } = render(<ButtonGroup id='test' />)
  expect(container.querySelector('div').getAttribute('id')).toEqual('test')
})
