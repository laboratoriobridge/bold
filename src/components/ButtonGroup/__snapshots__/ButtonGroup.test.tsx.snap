// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-1 > button:not(:first-of-type),
.emotion-1 > a:not(:first-of-type) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.emotion-1 > button:not(:last-of-type),
.emotion-1 > a:not(:last-of-type) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.emotion-1 > button:focus,
.emotion-1 > a:focus {
  z-index: 1;
}

.emotion-0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-size: 0.8125rem;
  line-height: 1.5rem;
  color: #3A3A47;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  font-weight: bold;
  cursor: pointer;
  background-color: #FFFFFF;
  border: 1px solid #696979;
  border-radius: 4px;
  color: #3A3A47;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 0.875rem;
  padding: calc(0.75rem - 1px) 2.5rem;
}

.emotion-0 > span {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0:not(:disabled):active {
  box-shadow: inset 0 2px 8px 0 rgba(0,0,0,0.1);
}

.emotion-0:focus {
  outline: none;
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-0:not(:disabled):hover {
  background-color: #F0F0F5;
}

<div>
  <div
    class="emotion-1"
    role="group"
  >
    <button
      class="emotion-0"
      type="button"
    >
      <span>
        First
      </span>
    </button>
  </div>
</div>
`;

exports[`should should accept style prop 1`] = `
.emotion-0 {
  color: red;
}

.emotion-0 > button:not(:first-of-type),
.emotion-0 > a:not(:first-of-type) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.emotion-0 > button:not(:last-of-type),
.emotion-0 > a:not(:last-of-type) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.emotion-0 > button:focus,
.emotion-0 > a:focus {
  z-index: 1;
}

<div>
  <div
    class="emotion-0"
    role="group"
  />
</div>
`;
